vetomint/tests/integration_test.rs

```diff
@@ -83,3 +85,582 @@ fn success_trivial_1() {
+        };
+        let response = progress(&height_info, &mut state, event);
+        assert!(response.is_empty());
+    }
+    let event = ConsensusEvent::Prevote {
+        proposal: 0,
+        round: 0,
+        signer: 3,
+        time: 3,
+    };
+
+    let response = progress(&height_info, &mut state, event);
+    assert_eq!(
+        response,
+        vec![ConsensusResponse::BroadcastPrecommit {
+            proposal: 0,
+            round: 0
+        }]
+    );
+
+    // STEP 3: Precommit.
+    for validator_index in 0..=2 {
+        let event = ConsensusEvent::Precommit {
+            proposal: 0,
+            round: 0,
+            signer: validator_index,
+            time: 4,
+        };
+        let response = progress(&height_info, &mut state, event);
+        assert!(response.is_empty());
+    }
+    let event = ConsensusEvent::Precommit {
+        proposal: 0,
+        round: 0,
+        signer: 3,
+        time: 4,
+    };
+    let response = progress(&height_info, &mut state, event);
+    assert_eq!(
+        response,
+        vec![ConsensusResponse::FinalizeBlock { proposal: 0 }]
+    );
+}
+
+#[ignore]
+#[test]
+// New round with 2/3+ prevote e.t. and nil committed due to timeout.
+fn early_termination_and_timeout() {
+    let height_info = HeightInfo {
+        validators: vec![1, 1, 1, 1, 1, 1, 1],
+        this_node_index: 6,
+        timestamp: 0,
+        consensus_params: ConsensusParams { timeout_ms: 1000 },
+    };
+    let mut state = ConsensusState::new(height_info.clone());
+
+    // STEP 1: Proposal.
+    let event = ConsensusEvent::BlockProposal {
+        proposal: 0,
+        proposer: 0,
+        round: 0,
+        time: 1,
+    };
+    let response = progress(&height_info, &mut state, event);
+    assert!(response.is_empty());
+    let event = ConsensusEvent::ProposalFavor {
+        proposal: 0,
+        favor: true,
+        time: 2,
+    };
+    let response = progress(&height_info, &mut state, event);
+    assert_eq!(
+        response,
+        vec![ConsensusResponse::BroadcastPrevote {
+            proposal: 0,
+            round: 0
+        }]
+    );
+
+    // STEP 2: Prevote.
+    for validator_index in 0..=2 {
+        let event = ConsensusEvent::Prevote {
+            proposal: 0,
+            round: 0,
+            signer: validator_index,
+            time: 3,
+        };
+        let response = progress(&height_info, &mut state, event);
+        assert!(response.is_empty());
+    }
+    let event = ConsensusEvent::Prevote {
+        proposal: 0,
+        round: 0,
+        signer: 3,
+        time: 3,
+    };
+    let response = progress(&height_info, &mut state, event);
+    assert_eq!(
+        response,
+        vec![ConsensusResponse::BroadcastPrecommit {
+            proposal: 0,
+            round: 0
+        }]
+    );
+
+    let timeout = height_info.consensus_params.timeout_ms as i64;
+
+    //Step3: Timeout
+    let event = ConsensusEvent::Timer { time: timeout };
```

Comment:
this will not trigger timeout,
because prevote phase is started at time 3, not 0